src/cargo/core/compiler/build_runner/mod.rs

```diff
@@ -422,6 +426,21 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
         self.files().metadata(unit)
     }
 
+    /// Returns the list of SBOM output file paths for a given Unit.
+    ///
+    /// Only call this function when `sbom` is active.
+    pub fn sbom_output_files(&self, unit: &Unit) -> CargoResult<Vec<PathBuf>> {
+        assert!(self.bcx.build_config.sbom);
+        let files = self
+            .outputs(unit)?
+            .iter()
+            .filter(|o| matches!(o.flavor, FileFlavor::Normal | FileFlavor::Linkable))
+            .filter_map(|output_file| output_file.hardlink.as_ref())
+            .map(|link_dst| link_dst.with_extension("cargo-sbom.json"))
```

Comment:
Do we want to handle name collisions here? For example dylibs might collide with each their as their name have no `-<hash>` suffix.

See <https://github.com/rust-lang/cargo/issues/6313> for some details.

Reply:
Tbh, I'm not sure how to solve this.

Reply:
This is already a problem for debug info, e.g. .pdb files on Windows. They have to be attributable to the file they correspond to, so they have the same name and cause a name collision. Cargo currently prints a warning about it, but otherwise just lets the collision happen and one of the debug info files is lost. However, losing SBOM info is highly undesirable.

We are not constrained by having the exact same name for the SBOM (unlike for debug info where we don't control the way it is discoverd), so we could do something about this.

I think we should emit all SBOM files in case of a collision, with a predictable naming scheme such as adding a number to the extension - `foo.sbom`, then `foo.1.sbom` and so on. The SBOM file will then contain the name of the binary it corresponds to, so that it can be disambiguated.

Reply:
The SBOM file extension is appended, the existing file extension is not replaced.